conda_smithy/azure_ci_utils.py

```diff
@@ -0,0 +1,191 @@
+import os
+import typing
+
+from vsts.vss_connection import VssConnection
+from msrest.authentication import BasicAuthentication
+
+
+AZURE_TEAM_INSTANCE = os.getenv("AZURE_INSTANCE", "https://dev.azure.com/conda-forge")
+AZURE_PROJECT_ID = os.getenv("AZURE_PROJECT_ID", "feedstock-builds")
+AZURE_SERVICE_ENDPOINT_NAME = os.getenv("AZURE_SERVICE_ENDPOINT", "conda-forge")
+
+try:
+    # Create a token at https://circleci.com/account/api. Put it in circle.token
+    with open(os.path.expanduser("~/.conda-smithy/azure.token"), "r") as fh:
+        AZURE_TOKEN = fh.read().strip()
+    if not AZURE_TOKEN:
+        raise ValueError()
+except (IOError, ValueError):
+    print(
+        "No azure token.  Create a token at https://dev.azure.com/conda-forge/_usersSettings/tokens and\n"
+        "put it in ~/.conda-smithy/azure.token"
+    )
+
+
+credentials = BasicAuthentication("", AZURE_TOKEN)
+connection = VssConnection(base_url=AZURE_TEAM_INSTANCE, creds=credentials)
+
+
+def get_service_endpoint(project_id=AZURE_PROJECT_ID):
+    from vsts.service_endpoint.v4_1.service_endpoint_client import ServiceEndpointClient
+    from vsts.service_endpoint.v4_1.models import ServiceEndpoint
+
+    service_endpoint_client = ServiceEndpointClient(
+        base_url=AZURE_TEAM_INSTANCE, creds=credentials
+    )
+    endpoints: typing.List[
+        ServiceEndpoint
+    ] = service_endpoint_client.get_service_endpoints(project=project_id, type="GitHub")
+    for service_endpoint in endpoints:
+        if service_endpoint.name == AZURE_SERVICE_ENDPOINT_NAME:
+            return service_endpoint
+    else:
+        raise KeyError("Service endpoint not found")
+
+
+def get_queues(project_id=AZURE_PROJECT_ID):
+    from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
+    from vsts.task_agent.v4_0.models import TaskAgentQueue
+
+    aclient = TaskAgentClient(AZURE_TEAM_INSTANCE, credentials)
+    queues: typing.List[TaskAgentQueue] = aclient.get_agent_queues(project_id)
+    return queues
+
+
+def get_default_queue(project_id):
+    queues = get_queues(project_id)
+    for q in queues:
+        if q.name == "Default":
+            return q
+    else:
+        raise ValueError("Default queue not found")
+
+
+def get_repo_reference(project_id, github_org, repo_name):
+    from vsts.build.v4_1.build_client import BuildClient
+    from vsts.build.v4_1.models import SourceRepositories, SourceRepository
+
+    service_endpoint = get_service_endpoint(project_id)
+    bclient: BuildClient = connection.get_client(
```

Comment:
Well, I think I found a bug in xonsh :)

Reply:
Just checking, was this resolved and/or is there an issue/PR we should xref to?